# 문제1

In [ ]:
1. 타이타닉호에 탄 클래스별 남여의 평균나이를 구하시오
2. 지역별 남여 생존자 수를 구하시오

In [298]:
import pandas as pd

In [299]:
import seaborn as sns

In [300]:
df = sns.load_dataset('titanic')

In [301]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [305]:
pdf1 = df.pivot_table(index = 'pclass',    # 행 위치에 들어갈 열
                      columns = 'sex',    # 열 위치에 들어갈 열
                      values = 'age',     # 데이터로 사용할 열
                      aggfunc = 'mean')   # 데이터 집계함수
pdf1

sex        female       male
pclass                      
1       34.611765  41.281386
2       28.722973  30.740707
3       21.750000  26.507589

In [308]:
pdf2 = df.pivot_table(index='embark_town', 
                      columns='sex',
                      values='survived',
                      aggfunc=['sum','mean'])
pdf2

sum           mean          
sex         female male    female      male
embark_town                                
Cherbourg       64   29  0.876712  0.305263
Queenstown      27    3  0.750000  0.073171
Southampton    140   77  0.689655  0.174603

#### 결론:Southampton 여성은 140명 살았고 생존율은68.9%이다.

# 문제2

#### 1.클래스등급과 남여그룹별 나이,요금,생존율 평균을 구하시오.
#### 2.조회된 자료에서 여성만 추출하여 보여주시오.  

In [315]:
df = sns.load_dataset('titanic')[['age','sex','class','fare','survived']]
gdf = df.groupby(['class','sex']).mean()
gdf

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
       male    41.281386   67.226127  0.368852
Second female  28.722973   21.970121  0.921053
       male    30.740707   19.741782  0.157407
Third  female  21.750000   16.118810  0.500000
       male    26.507589   12.661633  0.135447

In [320]:
gdf1 = gdf.query('sex == "female"')
gdf1

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
Second female  28.722973   21.970121  0.921053
Third  female  21.750000   16.118810  0.500000

In [326]:
gds2 =  gdf1.index.get_level_values(0)
gds2

CategoricalIndex(['First', 'Second', 'Third'], categories=['First', 'Second', 'Third'], ordered=False, dtype='category', name='class')

#  문제3

### 이름으로 Inner Join,Left Outer/Right Outer/Outer Join을 하시오

In [427]:
from pandas import Series, DataFrame

In [428]:
JDF1 = DataFrame([["홍길동",180,75,1.5],["이순신",160,65,2.0],["이춘복",170,75,0.5]], columns = ["이름","키","몸무게","시력"])
JDF1

이름    키  몸무게   시력
0  홍길동  180   75  1.5
1  이순신  160   65  2.0
2  이춘복  170   75  0.5

In [429]:
JDF2 = DataFrame([["홍길동","포워드","잘함","포악함"],["이순신","미드필더","못함","온순함"],["맹꽁이","수비수","잘함","화잘냄"]], columns = ["이름","포지션","실력","성격"])
JDF2

이름   포지션  실력   성격
0  홍길동   포워드  잘함  포악함
1  이순신  미드필더  못함  온순함
2  맹꽁이   수비수  잘함  화잘냄

In [430]:
pd.merge(left = JDF1 , right = JDF2, how = "inner", on = "이름")

이름    키  몸무게   시력   포지션  실력   성격
0  홍길동  180   75  1.5   포워드  잘함  포악함
1  이순신  160   65  2.0  미드필더  못함  온순함

In [431]:
pd.merge(left = JDF1 , right = JDF2, how = "left", on = "이름")

이름    키  몸무게   시력   포지션   실력   성격
0  홍길동  180   75  1.5   포워드   잘함  포악함
1  이순신  160   65  2.0  미드필더   못함  온순함
2  이춘복  170   75  0.5   NaN  NaN  NaN

In [432]:
pd.merge(left = JDF1 , right = JDF2, how = "right", on = "이름")

이름      키   몸무게   시력   포지션  실력   성격
0  홍길동  180.0  75.0  1.5   포워드  잘함  포악함
1  이순신  160.0  65.0  2.0  미드필더  못함  온순함
2  맹꽁이    NaN   NaN  NaN   수비수  잘함  화잘냄

In [433]:
pd.merge(left = JDF1 , right = JDF2, how = "outer", on = "이름")

이름      키   몸무게   시력   포지션   실력   성격
0  홍길동  180.0  75.0  1.5   포워드   잘함  포악함
1  이순신  160.0  65.0  2.0  미드필더   못함  온순함
2  이춘복  170.0  75.0  0.5   NaN  NaN  NaN
3  맹꽁이    NaN   NaN  NaN   수비수   잘함  화잘냄

## 참고. merge()함수사용 병합하기

In [439]:
df1 = pd.DataFrame({
    '고객번호': [1, 2, 3, 4, 5, 6, 7],
        '이름': ['가희', '나희', '다희', '라희', '마희', '사희', '아희']
}, columns=['고객번호', '이름'])
df1

고객번호  이름
0     1  가희
1     2  나희
2     3  다희
3     4  라희
4     5  마희
5     6  사희
6     7  아희

In [440]:
df2 = pd.DataFrame({
    '고객번호': [1, 1, 5, 6, 8, 10],
        '금액': [10000, 20000, 15000, 5000, 100000, 30000]
}, columns=['고객번호', '금액'])
df2

고객번호      금액
0     1   10000
1     1   20000
2     5   15000
3     6    5000
4     8  100000
5    10   30000

In [441]:
pd.merge(df1, df2)

고객번호  이름     금액
0     1  가희  10000
1     1  가희  20000
2     5  마희  15000
3     6  사희   5000

## 문제4

### df1 자료의 1행의 데이터를 추가해서 df1에 df2를 합시치오

In [448]:
import pandas as pd

In [449]:
import numpy as np

In [457]:
columns = np.array(['서울','경기','전라','제주'])

In [458]:
data = np.array([['서1','경1','전1','제1'],
                 ['서2','경2','전2','제2'],
                 ['서3','경3','전3','제3'],
                 ['서4','경4','전4','제4']])

In [473]:
df1 = pd.DataFrame(data,columns=columns)

In [474]:
df1

서울  경기  전라  경상
0  서1  경1  전1  경1
1  서2  경2  전2  경2
2  서3  경3  전3  경3
3  서4  경4  전4  경4

In [470]:
data1 = np.array([['서11','경11','전11','제11'],
                  ['서12','경12','전12','제12'],
                  ['서13','경13','전13','제13'],
                  ['서14','경14','전14','제14']])

In [475]:
df2 = pd.DataFrame(data1,columns=columns)

In [476]:
df2

서울   경기   전라   경상
0  서11  경11  전11  경11
1  서12  경12  전12  경12
2  서13  경13  전13  경13
3  서14  경14  전14  경14

In [477]:
df_con = pd.concat([df1,df2])

In [478]:
df_con

서울   경기   전라   경상
0   서1   경1   전1   경1
1   서2   경2   전2   경2
2   서3   경3   전3   경3
3   서4   경4   전4   경4
0  서11  경11  전11  경11
1  서12  경12  전12  경12
2  서13  경13  전13  경13
3  서14  경14  전14  경14

In [481]:
df1.loc[4] = ['서21','경21','전21','제21']

In [483]:
df1

서울   경기   전라   경상
0   서1   경1   전1   경1
1   서2   경2   전2   경2
2   서3   경3   전3   경3
3   서4   경4   전4   경4
5  서21  경21  전21  경21

In [484]:
df1.append(df2)

C:\Users\TJ\AppData\Local\Temp\ipykernel_7172\3062608662.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df1.append(df2)


서울   경기   전라   경상
0   서1   경1   전1   경1
1   서2   경2   전2   경2
2   서3   경3   전3   경3
3   서4   경4   전4   경4
5  서21  경21  전21  경21
0  서11  경11  전11  경11
1  서12  경12  전12  경12
2  서13  경13  전13  경13
3  서14  경14  전14  경14